### ISSUE TITLE
innerText implementation

### ISSUE THREAD

**Issue by vsemozhetbyt** [tmpvar/jsdom#1245]
*jsdom* is a great tool for web scraping. However the `textContent` is a very inconvenient way to get readable text for html2text conversion.

Tere is a wonderful article about usefulness of the negligible `innerText` in many cases:

http://perfectionkills.com/the-poor-misunderstood-innerText/

The author suggests `getSelection().toString()` as a very slow workaround, but `getSelection` is not implemented in the *jsdom* yet.

Could you consider an implementing of the `innerText` in the *jsdom*? The author has done a great exploration about it, he has even add a simple spec at the end.

**Comment by vsemozhetbyt** [comment]
And  what a pity that **rangy** `Selection` library is not compatible with **jsdom**: https://github.com/timdown/rangy/issues/348

**Comment by domenic** [comment]
So, innerText is not standard, and not implemented in at least one major engine (Firefox). Without a standard, I don't think we should implement it.

**Comment by Sebmaster** [comment]
Looks like there's some movement in this whole thing with a draft spec [here](https://rocallahan.github.io/innerText-spec/). See also all the [references](https://github.com/rocallahan/innerText-spec/#references). There are no issues on the repo though, so I wonder how complete it already is / how quick progress will be.

**Comment by vsemozhetbyt** [comment]
Firefox has implemented: https://bugzilla.mozilla.org/show_bug.cgi?id=264412

WHATWG semms to approve: https://github.com/whatwg/compat/issues/5#issuecomment-168049752

**Comment by inikulin** [comment]
From the spec it's seems like we can't implement `innerText` properly without basic layout support.

**Comment by domenic** [comment]
Yeah, this is not really going to be implementable in jsdom anyway, without a lot of infrastructure work... nobody get their hopes up :(.

**Comment by vsemozhetbyt** [comment]
As to layout support requirement: https://github.com/rocallahan/innerText-spec/issues/2

**Comment by vsemozhetbyt** [comment]
Is there any plan to implement it because of WHATWG adoption?

**Comment by domenic** [comment]
Yeah... Although the spec requires a lot of stuff jsdom doesn't have, around CSS boxes :(. Not sure what to do.

**Comment by vsemozhetbyt** [comment]
Is there any lib for this to plug along with jsdom?

**Comment by snuggs** [comment]
@domenic care to drop some knowledge on why this is such an infrastructure overhaul? We thought the 800lb gorilla in the room would leave lo-key. But looks like it's not going anywhere. As you know have been wrapping my head around the innards of jsdom. Where would be a great place in the repo to start reviewing code to a jsdom newb? 

Thanks in advance 🙏  /cc @vsemozhetbyt

**Comment by dmethvin** [comment]
The primary issue is the fact that `innerText` leans on the layout engine for guidance, and jsdom has no layout engine. See https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute and 
http://perfectionkills.com/the-poor-misunderstood-innerText/ . From the second link:

> Notice how innerText almost precisely represents exactly how text appears on the page. textContent, on the other hand, does something strange — it ignores newlines created by <br> and around styled-as-block elements (<span> in this case). But it preserves spaces as they are defined in the markup.

**Comment by vsemozhetbyt** [comment]
Still out of scope and no workaround?

**Comment by coreh** [comment]
Apparently the spec says: 

> If this element is not being rendered, **or if the user agent is a non-CSS user agent,** [emphasis added] then return the same value as the textContent IDL attribute on this element.

I think a workaround would be then to simply return `textContent`.

**Comment by domenic** [comment]
We implement enough CSS that I don't think that applies. We just don't implement the layout parts...

**Comment by Suzii** [comment]
Hi guys, any news on this one?

**Comment by Bnaya** [comment]
Just use headless chrome :)

**Comment by Janpot** [comment]
@domenic from that spec that @coreh mentioned:
https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute
>If this element is not being rendered, or if the user agent is a non-CSS user agent, then return the same value as the textContent IDL attribute on this element.

https://html.spec.whatwg.org/multipage/rendering.html#being-rendered
> An element is being rendered if it has any associated CSS layout boxes, SVG layout boxes, or some equivalent in other styling languages.

If jsdom doesn't implement the layout parts, doesn't that mean "not being rendered" applies?

**Comment by bennypowers** [comment]
This message is for anyone reaching this github thread that just wants a way to get their tests passing without changing their function implementations.

copypasta for the top of your test files:
```js
global.Element = (new JSDOM()).window.Element;
Object.defineProperty(global.Element.prototype, 'innerText', {
  get() {
    return this.innerHTML;
  },
});
```

Naturally, caveats from the above discussion apply.

**Comment by jettlin** [comment]
In case anyone else is running into this issue I took it 1 step further and used the `sanitize-html` package to get basically what the browser is doing (note I did not import the JSDOM setup as I found it wasn't needed when putting this in my Jest setup file but if you're not using Jest then you'll want to use the `global.Element = (new JSDOM()).window.Element` setup that @bennypowers recommended):

```javascript
Object.defineProperty(global.Element.prototype, 'innerText', {
  get() {
    return sanitizeHtml(this.textContent, {
      allowedTags: [], // remove all tags and return text content only
      allowedAttributes: {}, // remove all tags and return text content only
    });
  },
  configurable: true, // make it so that it doesn't blow chunks on re-running tests with things like --watch
});
```

**Comment by PierreTurnbull** [comment]
> In case anyone else is running into this issue I took it 1 step further and used the `sanitize-html` package to get basically what the browser is doing (note I did not import the JSDOM setup as I found it wasn't needed when putting this in my Jest setup file but if you're not using Jest then you'll want to use the `global.Element = (new JSDOM()).window.Element` setup that @bennypowers recommended):
> 
> ```js
> Object.defineProperty(global.Element.prototype, 'innerText', {
>   get() {
>     return sanitizeHtml(this.textContent, {
>       allowedTags: [], // remove all tags and return text content only
>       allowedAttributes: {}, // remove all tags and return text content only
>     });
>   },
>   configurable: true, // make it so that it doesn't blow chunks on re-running tests with things like --watch
> });
> ```

The newly created `innerText` displays strings that are in a `display: none` div, whereas the native `innerText` doesn't.

**Comment by macgyver** [comment]
i had a similar need but wanted to go slightly further than just using the `textContent` - again, this won't be an accurate representation of what browsers actually do, especially with respect to visibly hidden elements, but it's good enough for my use case: 
```
function innerText(el)
  el.querySelectorAll('script,style').forEach(el => el.remove())
  return el.textContent
```

**Comment by ty5491003** [comment]
What a pity！

**Comment by milahu** [comment]
>> Apparently the spec says:
>>>    If this element is not being rendered, **or if the user agent is a non-CSS user agent**, [emphasis added] then return the same value as the textContent IDL attribute on this element.
>>
>> I think a workaround would be then to simply return textContent.
>
> We implement enough CSS that I don't think that applies. We just don't implement the layout parts...

@domenic please consider a more liberal interpretation of the spec

`textContent` is explicitly allowed as fallback, when application of CSS rules is too expensive

**Comment by domenic** [comment]
Given that I am the spec editor, I can state with certainty that "when application of CSS rules is too expensive" is not what the spec is saying.

**Comment by milahu** [comment]
.. that was my interpretation of "if the user agent is a non-CSS user agent"

whats the difference between a "CSS user agent" and a "non-CSS user agent"?

what about:
a CSS user agent can "apply CSS rules" and output the result (graphic or textual)
a non-CSS user agent is too dumb to "apply CSS rules"

> We implement enough CSS that I don't think that applies.

what do you mean? window.getComputedStyle?

a fallback to textContent is still better than not implementing a standard interface

**Comment by aprilandjan** [comment]
Maybe we can just use `textContent` value  to replace the result of `innerText` while running tests with `jsdom`. For example:

```js
describe('mytest', () => {
  beforeAll(() => {
    Object.defineProperty(HTMLElement.prototype, 'innerText', {
      get() {
        return this.textContent;
      }
    });
  });
  it('should ok', () => {
  // test assertions
  });
});
```

**Comment by Shaleen-25** [comment]
```
    var element = document.createElement('p');
      element.setAttribute('class', 'data');
      element.innerText = 'data';
```

If you're facing issues while writing unit tests and getting undefined for innertext/outerText since jsdom doesn't support these is to specifically assign innertext/outerText to the dom element and then pass on this element.

**Comment by ellis** [comment]
For my use case, it is necessary to separate the text from different elements.  `innerText` tends to do this with newlines, whereas `textContent` just runs everything together.  Here is the function I ended up writing to return a string array with jsdom:

```
/**
 * Recursively iterate into the node, and return an array of the non-empty text of all text nodes.
 */
function extractTextArray(node: ChildNode): string[] {
	const textLines: string[] = [];
	function stepThru(node: ChildNode): void {
		for (let k = 0; k < node.childNodes.length; k++) {
			const x = node.childNodes[k];
			if (x.nodeType === node.TEXT_NODE) {
				const s = x.textContent?.trim();
				if (s) {
					textLines.push(s);
				}
			}
			stepThru(x);
		}
	}
	stepThru(node);
	return textLines;
}
```

**Comment by brusnom** [comment]
Getting text with `\n` by `textContent`...

```
function getTextAsInnerText(node) {
  const textLines = [];
  function stepThru(node) {
    for (let k = 0; k < node.childNodes.length; k++) {
      const x = node.childNodes[k];
      if (x.nodeType === node.TEXT_NODE) {
        textLines.push(x.textContent ? x.textContent.trim() : "");
      }
      stepThru(x);
    }
  }
  stepThru(node);
  return textLines.join('\\n');
}

Object.defineProperty(HTMLElement.prototype, 'innerText', {
  get() {
    return getTextAsInnerText(this)
  }
});
```
@ellis @aprilandjan thanks!

**Comment by kevduc** [comment]
I don't see any solution for this issue yet, so I'm adding my use case: I'm doing some web scrapping and I was testing a few things on a page directly in the chrome dev tools console. When things were working fine, I copy-pasted the code to node and got some error about trying to access a property of `undefined`.
I spent some time debugging to finally realize that getting `innerText` returns `undefined`, which was a bit frustrating and is something I didn't expect. I went online and eventually found this issue. In my case I can simply replace `innerText` by `textContent` which is fine.

@Janpot wrote:
> @domenic from that spec that @coreh mentioned: https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute
> 
> > If this element is not [being rendered](https://html.spec.whatwg.org/multipage/rendering.html#being-rendered), or if the user agent is a non-CSS user agent, then return the same value as the `textContent` IDL attribute on this element.
> 
> https://html.spec.whatwg.org/multipage/rendering.html#being-rendered
> 
> > An element is **being rendered** if it has any associated CSS layout boxes, SVG layout boxes, or some equivalent in other styling languages.
> 
> If `jsdom` doesn't implement the layout parts, doesn't that mean "not being rendered" applies?

From reading the previous discussions, I understand that it would be complicated to have to render layouts to figure out what `innerText` should contain, but like @Janpot, I'm not sure I understand why `innerText` doesn't return the same value as `textContent` in that case.

In any case, considering the number of dups and comments on this issue I feel like it could be good to mention this limitation somewhere, maybe directly in the readme? (and it would also appear on https://www.npmjs.com/package/jsdom)

**Comment by vozeldr** [comment]
Of course, I got here because I'm having this completely unexpected issue to. Maybe I'm not entirely understanding the discussion but it seems the push back boils down to 2 points: it's a lot of work, implementing it would not be compliant. It seems like an "all or nothing" mentality where instead of SOME implementation that would work for A LOT of cases would be simple, but you're completely disregarding that. Here's an example of entirely unexpected behavior:

```typescript
import {TextDecoder, TextEncoder} from 'util';

global.TextEncoder = TextEncoder;
// @ts-ignore
global.TextDecoder = TextDecoder;
const jsdom = require('jsdom');
const {JSDOM} = jsdom;

class Example {
    constructor(private doc: Document) {}

    public setElementText(selector: string, text: string): void {
        const element = this.doc.querySelector(selector);
        if (element && element instanceof HTMLElement) {
            element.innerText = text;
        }
    }
}

describe('Example', () => {
    it('should set the text', () => {
        const dom = new JSDOM(`
                <!DOCTYPE html>
                <html lang="en">
                    <body>
                        <span id="test">Test</span>
                    </body>
                </html>
            `);
        const example = new Example(dom.window.document);
        const span = dom.window.document.getElementById('test');
        example.setElementText('#test', 'confirmed');
        expect(span.innerText).toBe('confirmed');
    });
});
```

Yields the result:

```
> jest "Example"

 FAIL  test/options/example.spec.ts
  Example
    ✕ should set the text (46 ms)

  ● Example › should set the text

    expect(received).toBe(expected) // Object.is equality

    Expected: "confirmed"
    Received: undefined

      31 |         const span = dom.window.document.getElementById('test');
      32 |         example.setElementText('#test', 'confirmed');
    > 33 |         expect(span.innerText).toBe('confirmed');
         |                                ^
      34 |     });
      35 | });
      36 |
```

All the code I'm testing there is doing is setting the value of a property and then the test is verifying the value of the property was set... very simple stuff, but it seems like someone has actually gone out of their way to make this not work. Even if jsdom didn't define `innerText` on the object, when I set it I should then be able to read a value back but it seems like a getter/setter that intentionally breaks that simple capability was put in place instead.

**Comment by milahu** [comment]
abandon hope, all ye who enter here

@domenic is too busy with [hostage-taking](https://neilonsoftware.com/difficult-people-on-software-projects/developers/the-hostage-taker/) jsdom with his opinions
cos, you know, he is THE SPEC EDITOR (peace be upon him)

> > > > Apparently the spec says:
> > > > > If this element is not being rendered, **or if the user agent is a non-CSS user agent**, [emphasis added] then return the same value as the textContent IDL attribute on this element.
> > > > 
> > > > 
> > > > I think a workaround would be then to simply return textContent.
> > > 
> > > 
> > > We implement enough CSS that I don't think that applies. We just don't implement the layout parts...
> > 
> > @domenic please consider a more liberal interpretation of the spec
> > 
> > `textContent` is explicitly allowed as fallback, when application of CSS rules is too expensive
> 
> Given that I am the spec editor, I can state with certainty that "when application of CSS rules is too expensive" is not what the spec is saying.
> 

back to square one 🤡

this bug will never leave the main repo ... its a feature!

**Comment by amakhrov** [comment]
IMO a part of the problem is that the lack of innerText support is unexpected to a developer who just tries to write some tests for an innocently looking piece of code. The test fails, and then debugging starts - only to discover that there is nothing wrong with the code, but rather with the testing tool.

I can imagine this made smoother by throwing a custom exception explaining this limitation of jsdom. Perhaps including a possible workaround or a link to the docs.

The "partial" workaround (fall back to textContent) as suggested above might be sufficient for some use cases, but not other. It's ok to have it on the opt-in basis. Frankly, it was quite trivial to add it into my jest.setup.js - but before that I had to dig and figure out what exactly the limitations and tradeoffs are.

**Comment by aquelehugo** [comment]
@domenic Could we do something about this? I understand using `textContent` as fallback is not ideal and I guess you are also concerned about breaking changes, but what about simply talking about it in the documentation?

My suggestion is putting a section on README saying `innerText` is not implemented and linking to another document with details. In the linked document, the fallback snippet could be also added, with an explanation about the drawbacks. 

WDYT?

**Comment by AntonKhorev** [comment]
It could at least [fail with an exception](https://github.com/jsdom/jsdom/issues/1245#issuecomment-966545022) instead of quietly returning undefined.

**Comment by aquelehugo** [comment]
> It could at least [fail with an exception](https://github.com/jsdom/jsdom/issues/1245#issuecomment-966545022) instead of quietly returning undefined.

@AntonKhorev that makes sense indeed, but it would be a breaking change. Some people may not have `innerText` explicitly in their tests, but dependencies may use it internally. So after years of working tests, people would be obligated to make a decision about how to handle it on their own.
Maybe logging a warning instead of throwing an exception could be a middle ground.

Anyway, as you can see, this issue is more than 6 years long, so I'm hoping we can move it forward with baby steps

**Comment by jacopomazzoni** [comment]
> For my use case, it is necessary to separate the text from different elements. `innerText` tends to do this with newlines, whereas `textContent` just runs everything together. Here is the function I ended up writing to return a string array with jsdom:
> 
> ```
> /**
>  * Recursively iterate into the node, and return an array of the non-empty text of all text nodes.
>  */
> function extractTextArray(node: ChildNode): string[] {
> 	const textLines: string[] = [];
> 	function stepThru(node: ChildNode): void {
> 		for (let k = 0; k < node.childNodes.length; k++) {
> 			const x = node.childNodes[k];
> 			if (x.nodeType === node.TEXT_NODE) {
> 				const s = x.textContent?.trim();
> 				if (s) {
> 					textLines.push(s);
> 				}
> 			}
> 			stepThru(x);
> 		}
> 	}
> 	stepThru(node);
> 	return textLines;
> }
> ```

this would be great if it didn't change the order of the elements.

**Comment by RodolpheGohard** [comment]
Hello, I wrote a naive function that immitates the innerText behaviour:

```js
function naiveInnerText(node: Node) {
    const Node = node; // We need Node(DOM's Node) for the constants, but Node doesn't exist in the nodejs global space, and any Node instance references the constants through the prototype chain
    return [...node.childNodes].map(node => {
        switch (node.nodeType) {
            case Node.TEXT_NODE:
                return node.textContent;
            case Node.ELEMENT_NODE:
                return naiveInnerText(node);
            default:
                return "";
        }
    }).join("\n");
}
```

Please note that, contrarily to innerText, it prints out all descendants textNodes of given element, ignoring if they should be displayed or not in a browser. eg contents like `<script>`, `<head>`, or elements with `display:none` will be output by above function, when they are not if you do it from your browser implementation. I think you'd need the layout engine the contributors are mentionning above to figure that out.

PS: I think throwing an error when accessing .innerText would be the most sensible thing to do in JSDOM. Saves the confusion and time looking for an explanation.

**Comment by Nokel81** [comment]
So furthermore, though this might be easier to implement, setting `innerText` to `""` does not clear all children, as the spec days it should.

See the above two issues as references.

**Comment by rsov** [comment]
Came into an issue today when trying to use `innerText` in jest. Would have saved me some time if it threw "not implemented in jsdom" instead of `undefined`

**Comment by DavidTeju** [comment]
> It could at least [fail with an exception](https://github.com/jsdom/jsdom/issues/1245#issuecomment-966545022) instead of quietly returning undefined.

I very much agree with this. Just had a 30-minute to an hour-long frenzy trying to debug my code after getting undefined from innerText. Would've certainly saved me time and hassle to have just received an exception

**Comment by g-cheishvili** [comment]
> 

This is very annoying indeed!

**Comment by DavidTeju** [comment]
> So, innerText is not standard, and not implemented in at least one major engine (Firefox). Without a standard, I don't think we should implement it.

Bringing this to your attention again. This has been standard since 2016. Either implement it or throw an exception!! How hard could it be?

**Comment by TimothyGu** [comment]
@aquelehugo We have not unsubscribed from this thread.

@DavidTeju if you find `innerText` to be important to your use cases, we would definitely welcome any contribution that implements the current spec. None of us is paid to write these features, and it's unreasonable to expect any progress unless you are willing to do the work yourself.

**Comment by aquelehugo** [comment]
@TimothyGu If contributing with the implementation by spec is all it takes, I will work on it

**Comment by aquelehugo** [comment]
@DavidTeju nice! I was planning on doing that on the weekend, but I can try to push some code in my spare time through the week too. I can message you when I do.
Thanks for taking care of this :smile:

**Comment by DavidTeju** [comment]
Been very busy with school but the semester's over now so I'll get back on this sometime this week. Should have a working sample by next Monday.

**Comment by aaronfc** [comment]
As a workaround I have decided to use the [striptags package:](https://github.com/ericnorris/striptags) and it's working like a charm for me.

**Comment by DavidTeju** [comment]
@aaronfc I believe that would be the equivalent of Element.textContent as opposed to Element.innerText. Please correct me if I am wrong.
Prep for my internship has been taking much of my time so I'm way behind schedule on what I promised (Monday).

**Comment by aquelehugo** [comment]
@DavidTeju Pretty much. He could simply use `Element.textContent`, unless he is using the tag filtering options from `striptags`.

**Comment by bennbollay** [comment]
Any progress updates on this since mid-May?

**Comment by DavidTeju** [comment]
I made some good progress on this but then my internship started in late May and I’ve been super swamped with that. Planning to finish up in August, when my internship’s over and I’m back at school but if anyone wants to finish off, I can add some mental notes in comments and push to my fork. If not, I’ll just get to it when I can and submit a PR with the full changes.

**Comment by bennbollay** [comment]
Congratulations on the internship, best of luck!  Thank you for your continued effort here :)

**Comment by aquelehugo** [comment]
@DavidTeju I can take over. Please push the missing changes/notes to your fork

**Comment by Naragod** [comment]
It has been 8 years since this question was last posted. 

I have implemented my own `innerText` implementation that works for my purposes.

```
const flatten = (arr: any[]): any[] => {
  return arr
    .reduce((prev, curr) => {
      return prev.concat(Array.isArray(curr) ? flatten(curr) : sanitizeString(curr));
    }, [])
    .filter((item: any) => item != undefined && item != "");
};

const innerText = (el: Element): string[] => {
  const textArray: string[] = flatten(
    [...el.querySelectorAll(":not(script,style,.hidden)")].map((item) => {
      if (item.children.length == 0) return item.textContent || "";
      return innerText(item);
    }),
  );
  // get all unique inputs from flattened array of strings.
  return [...new Set(textArray)];
};
```

**Comment by stefansundin** [comment]
I am using Selenium/WebDriver, JSDOM, and [mozilla/readability](https://github.com/mozilla/readability) to scrape text from a website. My current solution to extract the `innerText` data is to load the output from readability back into the webdriver and then extract it that way. Not the most elegant option but so far it is the best one that I've found. Since I'm already using selenium it is not a huge code smell, but it would be nice if there was a better way.

```typescript
const url = 'https://example.com/';

// scrape the website using selenium
const driver = await new Builder().forBrowser(Browser.FIREFOX).build();
await driver.get(url);
const pageSource = await driver.getPageSource();

// load the source into readability to extract the text
const doc = new JSDOM(pageSource, { url });
const reader = new Readability(doc.window.document);
const article = reader.parse();

// use selenium to extract the innerText from the readability content
await driver.get(`data:text/html;charset=utf-8,${article.content}`);
const innerText = await driver.executeScript<string>(() => {
  return document.body.innerText;
});
```

**Comment by andrewcartwright1** [comment]
Not the nicest of way's to do it but I had some cases where sometimes HTMLElement would work and sometimes it wouldnt.. I couldnt figure out why, so i changed it to be even more hacky. I mocked it using jest spies.

innerText.ts
```
let spyGet;
let spySet;

export const createSpies = () => {
  Object.defineProperty(Object.prototype, 'innerText', {
    get: () => {},
    set: () => {},
    configurable: true
  });

  // eslint-disable-next-line @typescript-eslint/ban-types
  spyGet = jest.spyOn(Object.prototype, 'innerText' as keyof Object, 'get');
  spyGet.mockImplementation(function () {
    return this.textContent
      .split('\\n')
      .filter((text: string) => text && !text.match(/^\\s+$/))
      .map((text: string) => text.trim())
      .join('\\n');
  });

  // eslint-disable-next-line @typescript-eslint/ban-types
  spySet = jest.spyOn(Object.prototype, 'innerText' as keyof Object, 'set');
  spySet.mockImplementation(function (value) {
    this.textContent = value;
  });
};

export const resetSpies = () => {
  spyGet?.mockRestore();
  spySet?.mockRestore();
};
```

Then you can call it using `createSpies` and `resetSpies` in your jest tests.

**Comment by macgyver** [comment]
can we close this ticket? people seem pretty capable of / content to roll
their own solutions

**Comment by DanKaplanSES** [comment]
Can this please be documented as a limitation in the README?  Speaking of which, is there a link that lists all the JSDOM limitations? If so, can this be added to the README? If it's already there, it seems like it should be found on the ["Unimplemented parts" section](https://github.com/jsdom/jsdom#unimplemented-parts-of-the-web-platform).

**Comment by DanKaplanSES** [comment]
> > It could at least [fail with an exception](https://github.com/jsdom/jsdom/issues/1245#issuecomment-966545022) instead of quietly returning undefined.
> 
> @AntonKhorev that makes sense indeed, but it would be a breaking change. Some people may not have `innerText` explicitly in their tests, but dependencies may use it internally.

In my opinion, that's even scarier to think about. It's hard enough to troubleshoot when it's in your own project, imagine how hard it would be when you're degrees of separation away from the code that caused the problem. 

I'm not arguing for a breaking change, but at least:

1. Document it
2. Provide some configuration that warns developers when it's depended on
3. Provide some configuration that throws an error when it's depended on
4. Use "the next best thing" as a fake response, such as `textContent`. 

IMO, any one of these would improve the current situation.

**Comment by victorhahncastell** [comment]
> Can this please be documented as a limitation in the README?

Second that. I even believe attempting to access innerText should at least yield some kind of warning. Now that innerText is standard this is just a horrible trap causing tests to fail in the most unpredictable kind of ways. Just spent hours trying to figure out what's wrong with my code. :'D

**Comment by dataf3l** [comment]
innertext is not in the standard? come on guys, you can do better than that. I mean seriously.

**Comment by aurium** [comment]
Hey @dataf3l, take it easy...

> innertext is not in the standard? come on guys, you can do better than that. I mean seriously.

`innerText` is not that simple to implement. Consider that:

```html
<p dir="auto">
    innertext is
    <del>not in the standard? come on guys, you can do better than that. I mean</del>
    seriously.
</p>
```

Now, see what happens when you just style the `<del>` to hide it from user, without removing it from the DOM:

```js
$0.innerText
// "innertext is not in the standard? come on guys, you can do better than that. I mean seriously."

$0.firstElementChild.style.display = 'none'

$0.innerText
// "innertext is seriously." 

$0.textContent
// "innertext is not in the standard? come on guys, you can do better than that. I mean seriously." 
```

Now you may understand why there are `textContent` and not `innerText` in a **DOM simulation** lib.

Now, if you dislike that lack of feature, you can write a patch or send a letter requesting your money back to "OPEN SOURCE SOLUTIONS CORPORATION, WALL STREET, NYC, NY".

**Comment by dataf3l** [comment]
I did take it easy I was going to write a much more aggressive message.
I am not happy until this is added.
I don't care it is not in the standard, outerHTML wasn't in the standard and still there it is anyway, so innerText is totally doable. it is not that I dislike the lack of the feature, I dislike the attitude towards the problem, saying "it's not in the standard I'm not doing it" is like come on man, we can do better than that.

**Comment by victorhahncastell** [comment]
While I fully appreciate @aurium's explanation -- and really thing innerText is one of those crazy features that should never have made it into the standard -- the lack of it in JSDOM is a massive trap.

I'd argue for aliasing innerText to textContent (as per @DanKaplanSES's suggestion no 4) as this will behave as expected in a large percentage of use cases. In fact, I was using innerText expecting it to behave the way textContent does because I simply did not know, and I'd assume it's the same for a large number of developers.

**Comment by JasonHK** [comment]
> I did take it easy I was going to write a much more aggressive message. I am not happy until this is added. I don't care it is not in the standard, outerHTML wasn't in the standard and still there it is anyway, so innerText is totally doable. it is not that I dislike the lack of the feature, I dislike the attitude towards the problem, saying "it's not in the standard I'm not doing it" is like come on man, we can do better than that.

@dataf3l You can submit a pull request. Like come on man, you can do better than that.

**Comment by arsinclair** [comment]
I just came across this issue after spending several hours trying to implement a custom text sanitization mechanism to replicate the missing innerText functionality. I wasn’t able to read the entire thread—it’s grown quite a bit over the years—but from what I gather, the absence of innerText in JSDOM is largely due to a strict interpretation of the spec, and possibly a reluctance to deviate from it.

That said, I wanted to express that this gap continues to cause a significant amount of pain for developers. I imagine countless engineers have spent a great deal of time, as I have, trying to build or find workarounds for this missing functionality. While I understand and respect the desire to adhere to standards, it’s frustrating when real-world developer needs are left unmet, especially for such a widely used and practically essential property.

**Comment by jacopomazzoni** [comment]
> I just came across this issue after spending several hours trying to implement a custom text sanitization mechanism to replicate the missing innerText functionality. I wasn’t able to read the entire thread—it’s grown quite a bit over the years—but from what I gather, the absence of innerText in JSDOM is largely due to a strict interpretation of the spec, and possibly a reluctance to deviate from it.
> 
> That said, I wanted to express that this gap continues to cause a significant amount of pain for developers. I imagine countless engineers have spent a great deal of time, as I have, trying to build or find workarounds for this missing functionality. While I understand and respect the desire to adhere to standards, it’s frustrating when real-world developer needs are left unmet, especially for such a widely used and practically essential property.

this was my exact experience and i still maintain that a simple error message would have gone a long way here

**Comment by maccman** [comment]
Can I give another +1 to support getting this implemented? I'm happy to pay for it.

**Comment by Stan-Stani** [comment]
An eslint override to help prevent us from getting bitten by this in the future:

(customize for your use-case)

```json
 "overrides": [
    {
      "files": ["**/*.test.{js,ts,jsx,tsx}"],
      "rules": {
        "no-restricted-properties": [
          "error",
          {
            "property": "innerText",
            "message": "Prefer the 'textContent' property. 'innerText' is always `undefined` in JSDOM. See https://github.com/jsdom/jsdom/issues/1245"
          }
        ]
      }
    }
  ],
```

**Comment by kirthi-b** [comment]
I'd like to try this per the current spec. The sticking point is always "being rendered", so agreeing on that first:

Most of the collection steps only need computed style, not boxes. display:none drops out, br is a newline, block boundaries add line breaks, table cells add tabs, visibility gates per node. My proposal: "being rendered" = connected with no display:none ancestor, run the full steps on computed style, validate against the innerText WPTs minus the soft-wrap cases that genuinely need layout.

If you'd draw the rendered line elsewhere, happy to adjust. Worth starting?

**Comment by domenic** [comment]
No, incomplete spec non-compliant implementations are not accepted by the jsdom project.

**Comment by Janpot** [comment]
> I'd like to try this per the current spec. The sticking point is always "being rendered", so agreeing on that first:

Is [this](https://html.spec.whatwg.org/multipage/dom.html#dom-innertext) the spec you're referring to?

> 1. If element is not [being rendered](https://html.spec.whatwg.org/multipage/rendering.html#being-rendered) or if the user agent is a non-CSS user agent, then return _element_'s [descendant text content](https://dom.spec.whatwg.org/#concept-descendant-text-content).

I'd say it's more the "non-CSS user agent" part that applies to jsdom, rather than the "not being rendered" part. But I can't find any definition in there of what they mean exactly by "css user-agent" vs. "non-css user-agent". It's pretty obvious to me personally under which category jsdom should fall, given that layout is out of scope for this project. But a true complete spec would indeed require a better definition of those terms 👍

**Comment by domenic** [comment]
The meaning of CSS user agent is pretty clear: it's a user agent that implements CSS. jsdom has a lot of CSS implementation, and so falls under that definition.

**Comment by victorhahncastell** [comment]
JSDOM is a great project and a very useful tool, but it's sad that for 11 years (!) now we keep discussing technicalities rather than find pragmatic solution. Any of the options suggested is better than the current one, which I might add is obviously not compliant to the spec either.

**Comment by Janpot** [comment]
> The meaning of CSS user agent is pretty clear: it's a user agent that implements CSS. jsdom has a lot of CSS implementation, and so falls under that definition.

Turns out the "being rendered" part is actually defined more strongly in [introduction](https://html.spec.whatwg.org/multipage/rendering.html#introduction-17):

> An element is **being rendered** if it has any associated CSS layout boxes, SVG layout boxes, or some equivalent in other styling languages.

jsdom does not do layout, therefor elements don't have associated layout boxes, therefor elements are not **being rendered**, therefor `innerText` should fall back to `textContent`, QED or am I missing something?

**Comment by DavidTeju** [comment]
At the very least, a not implemented exception is appropriate

**Comment by DanKaplanSES** [comment]
> At the very least, a not implemented exception is appropriate

This would be a breaking change for lots of old projects, and jsdom isn't always used as a direct dependency. It's a transitive dependency a lot of the time (e.g., something a test framework needs).  Therefore, upgrading a direct dependency and suddenly seeing this error would be very confusing for most.

That said, if jsdom had a linter or a way to opt in to a strict mode of some sort, I personally like your idea.

**Comment by arsinclair** [comment]
I am not sure an exception is the best idea here, so I'm only responding to the last comment: why are we afraid of breaking changes? When properly handled (major version bump, proper cooldown and freeze periods, migration guides) this shouldn't be a problem and we should be moving with confidence.
